### ISSUE TITLE
Implement QLoRA training pipeline

### ISSUE THREAD

**Issue by NP-compete** [NP-compete/smithery#8]
## Problem

The training pipeline is the second core pillar of smithery. Currently \`smithery/train/lora.py\` and \`smithery/train/callbacks.py\` raise \`NotImplementedError\`. We need a working QLoRA training loop that wraps PEFT + TRL with tool-calling-specific data formatting and callbacks.

### Training Loop (\`lora.py\`)

Wrap HuggingFace's PEFT (for LoRA) and TRL (for SFTTrainer) into a clean, opinionated pipeline:

```python
class AgentTrainer:
    def train(self, dataset):
        # 1. Load base model with 4-bit quantization (bitsandbytes)
        model = AutoModelForCausalLM.from_pretrained(
            self.config.base_model,
            quantization_config=BitsAndBytesConfig(
                load_in_4bit=True,
                bnb_4bit_quant_type="nf4",
                bnb_4bit_compute_dtype=torch.bfloat16,
            ),
        )
        
        # 2. Attach LoRA adapters
        peft_config = LoraConfig(
            r=self.config.lora.r,
            lora_alpha=self.config.lora.alpha,
            lora_dropout=self.config.lora.dropout,
            target_modules=self.config.lora.target_modules,
            task_type="CAUSAL_LM",
        )
        
        # 3. Format data using the right chat template
        # (auto-detect from base model or use config.data.format)
        
        # 4. Train with SFTTrainer
        trainer = SFTTrainer(
            model=model,
            train_dataset=formatted_dataset,
            peft_config=peft_config,
            args=TrainingArguments(...),
            callbacks=[cost_tracker, sample_generator, early_stopping],
        )
        trainer.train()
        
        # 5. Save adapter + return summary
```

### Key Design Decisions

1. **Auto-detect chat template**: Based on the base model name, automatically select the right format (ChatML for Phi-3/Qwen, Hermes for Mistral, Llama format for Llama). Override with \`config.data.format\`.

2. **Smart defaults**: The YAML configs should "just work" — no need to understand LoRA hyperparameters. Each pre-built recipe is tested and validated.

3. **Gradient checkpointing**: Always enabled for memory efficiency on consumer GPUs.

4. **Mixed precision**: bf16 on Ampere+, fp16 on older GPUs. Auto-detect.

### Callbacks (\`callbacks.py\`)

**CostTracker**: Estimate GPU cost based on utilization and cloud pricing.

**SampleGenerationCallback**: At configurable intervals during training, generate sample tool calls from a held-out prompt set and log them. This lets you visually verify the model is learning tool calling, not just memorizing.

**EarlyStoppingCallback**: Stop when eval loss plateaus. Patience and min_delta configurable.

### CLI Integration

```bash
smithery train \
  --config configs/phi3-tool-calling.yaml \
  --data ./data/train.jsonl \
  --output ./output/my-tool-agent
```

Should show a Rich progress bar with: current epoch, loss, learning rate, estimated cost, ETA.

## Acceptance Criteria

- [ ] \`AgentTrainer.train()\` produces a working LoRA adapter
- [ ] Supports all four pre-configured base models (Phi-3, Llama 3.2, Qwen 2.5, Mistral 7B)
- [ ] Auto-detects correct chat template from base model
- [ ] Gradient checkpointing enabled by default
- [ ] Mixed precision auto-detected (bf16 / fp16)
- [ ] CostTracker callback estimates GPU cost
- [ ] SampleGenerationCallback generates tool calls during training
- [ ] EarlyStoppingCallback stops on eval loss plateau
- [ ] Rich progress bar with loss, LR, cost, ETA
- [ ] \`TrainResult\` returned with loss, gpu_hours, cost, model_path
- [ ] Works on a single consumer GPU (16-24 GB VRAM)
- [ ] Unit tests (mocked model loading)
- [ ] Integration test on a tiny model (e.g., GPT-2) with synthetic data

## Dependencies

- Depends on #7 (formats.py for data preparation)
- Depends on \`SmitheryConfig\` and \`TrainResult\` types (implemented)

## References

- [PEFT QLoRA guide](https://huggingface.co/docs/peft/main/en/developer_guides/quantization)
- [TRL SFTTrainer](https://huggingface.co/docs/trl/main/en/sft_trainer)
- [Axolotl](https://github.com/axolotl-ai-cloud/axolotl) — Reference for training pipeline patterns

## Estimated Effort

Medium-Large (5-7 days). The PEFT/TRL integration is well-documented, but auto-detecting chat templates and the callback system require careful engineering.
